models/aeif_psc_delta.cpp

```diff
@@ -0,0 +1,533 @@
+
+nest::aeif_psc_delta::aeif_psc_delta( const aeif_psc_delta& n )
+  : Archiving_Node( n )
+  , P_( n.P_ )
+  , S_( n.S_ )
+  , B_( n.B_, *this )
+{
+}
+
+nest::aeif_psc_delta::~aeif_psc_delta()
+{
+  // GSL structs may not have been allocated, so we need to protect destruction
+  if ( B_.s_ )
+    gsl_odeiv_step_free( B_.s_ );
+  if ( B_.c_ )
+    gsl_odeiv_control_free( B_.c_ );
+  if ( B_.e_ )
+    gsl_odeiv_evolve_free( B_.e_ );
+}
+
+/* ----------------------------------------------------------------
+ * Node initialization functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::aeif_psc_delta::init_state_( const Node& proto )
+{
+  const aeif_psc_delta& pr = downcast< aeif_psc_delta >( proto );
+  S_ = pr.S_;
+}
+
+void
+nest::aeif_psc_delta::init_buffers_()
+{
+  B_.spikes_.clear();   // includes resize
+  B_.currents_.clear(); // includes resize
+  Archiving_Node::clear_history();
+
+  B_.logger_.reset();
+
+  B_.step_ = Time::get_resolution().get_ms();
+
+  // We must integrate this model with high-precision to obtain decent results
+  B_.IntegrationStep_ = std::min( 0.01, B_.step_ );
+
+  if ( B_.s_ == 0 )
+    B_.s_ =
+      gsl_odeiv_step_alloc( gsl_odeiv_step_rkf45, State_::STATE_VEC_SIZE );
+  else
+    gsl_odeiv_step_reset( B_.s_ );
+
+  if ( B_.c_ == 0 )
+    B_.c_ = gsl_odeiv_control_yp_new( P_.gsl_error_tol, P_.gsl_error_tol );
+  else
+    gsl_odeiv_control_init(
+      B_.c_, P_.gsl_error_tol, P_.gsl_error_tol, 0.0, 1.0 );
+
+  if ( B_.e_ == 0 )
+    B_.e_ = gsl_odeiv_evolve_alloc( State_::STATE_VEC_SIZE );
+  else
+    gsl_odeiv_evolve_reset( B_.e_ );
+
+  B_.sys_.jacobian = NULL;
+  B_.sys_.dimension = State_::STATE_VEC_SIZE;
+  B_.sys_.params = reinterpret_cast< void* >( this );
+  B_.sys_.function = aeif_psc_delta_dynamics;
+
+  B_.I_stim_ = 0.0;
+}
+
+void
+nest::aeif_psc_delta::calibrate()
+{
+  // ensures initialization in case mm connected after Simulate
+  B_.logger_.init();
+
+  // set the right threshold and GSL function depending on Delta_T
+  if ( P_.Delta_T > 0. )
+  {
+    V_.V_peak_ = P_.V_peak_;
+  }
+  else
+  {
+    V_.V_peak_ = P_.V_th; // same as IAF dynamics for spikes if Delta_T == 0.
+  }
+
+  V_.refractory_counts_ = Time( Time::ms( P_.t_ref_ ) ).get_steps();
+  // since t_ref_ >= 0, this can only fail in error
+  assert( V_.refractory_counts_ >= 0 );
+}
+
+/* ----------------------------------------------------------------
+ * Update and spike handling functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::aeif_psc_delta::update( const Time& origin,
+  const long from,
+  const long to )
+{
+  assert(
+    to >= 0 && ( delay ) from < kernel().connection_manager.get_min_delay() );
+  assert( from < to );
+  assert( State_::V_M == 0 );
+  const double h = Time::get_resolution().get_ms();
+  const double tau_m_ = P_.C_m / P_.g_L;
+  for ( long lag = from; lag < to; ++lag )
+  {
+    double t = 0.0;
+
+    if ( S_.r_ > 0 )
```

Comment:
See [my suggestion for a correct implementation](https://github.com/nest/nest-simulator/issues/547#issuecomment-262213725) in #547.